lib/src/rules/use_late_for_private_fields_and_variables.dart

```diff
@@ -0,0 +1,179 @@
+avoids null checks.
+
+**BAD:**
+```
+int? _i;
+m() {
+  _i!.abs();
+}
+```
+
+**GOOD:**
+```
+late int _i;
+m() {
+  _i.abs();
+}
+```
+
+''';
+
+class UseLateForPrivateFieldsAndVariables extends LintRule
+    implements NodeLintRule {
+  UseLateForPrivateFieldsAndVariables()
+      : super(
+          name: 'use_late_for_private_fields_and_variables',
+          description: _desc,
+          details: _details,
+          maturity: Maturity.experimental,
+          group: Group.style,
+        );
+
+  @override
+  void registerNodeProcessors(NodeLintRegistry registry,
+      [LinterContext context]) {
+    final visitor = _Visitor(this, context);
+    registry.addCompilationUnit(this, visitor);
+  }
+}
+
+class _Visitor extends UnifyingAstVisitor<void> {
+  _Visitor(this.rule, this.context);
+
+  final LintRule rule;
+  final LinterContext context;
+
+  CompilationUnitElement currentUnit;
+
+  static final lateables =
+      <CompilationUnitElement, List<VariableDeclaration>>{};
+  static final nullableAccess = <CompilationUnitElement, Set<Element>>{};
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    if (node.featureSet.isEnabled(Feature.non_nullable)) {
+      currentUnit = node.declaredElement;
+      lateables.putIfAbsent(currentUnit, () => []);
+      nullableAccess.putIfAbsent(currentUnit, () => {});
+
+      super.visitCompilationUnit(node);
+
+      final unitsInContext =
+          context.allUnits.map((e) => e.unit.declaredElement).toSet();
+      final libraryUnitsInContext = node.declaredElement.library.units
+          .where(unitsInContext.contains)
+          .toSet();
+      final areAllLibraryUnitsVisited =
+          libraryUnitsInContext.every(lateables.containsKey);
+      if (areAllLibraryUnitsVisited) {
+        _checkAccess(libraryUnitsInContext);
+
+        // clean up
+        libraryUnitsInContext.forEach((unit) {
+          lateables.remove(unit);
+          nullableAccess.remove(unit);
+        });
+      }
+    }
+  }
+
+  @override
+  void visitNode(AstNode node) {
+    var element = DartTypeUtilities.getCanonicalElementFromIdentifier(node);
+    if (element != null) {
+      var parent = node.parent;
+      if (parent is Expression) {
+        parent = (parent as Expression).unParenthesized;
+      }
+      if (node is SimpleIdentifier && node.inDeclarationContext()) {
+        // ok
+      } else if (parent is PostfixExpression &&
+          parent.operand == node &&
+          parent.operator.type == TokenType.BANG) {
+        // ok non-null access
+      } else if (parent is AssignmentExpression &&
+          parent.operator.type == TokenType.EQ &&
+          context.typeSystem.isNonNullable(parent.rightHandSide.staticType)) {
+        // ok non-null access
+      } else {
+        nullableAccess[currentUnit].add(element);
+      }
+    }
+    super.visitNode(node);
+  }
+
+  @override
+  void visitFieldDeclaration(FieldDeclaration node) {
```

Comment:
I believe that this needs to invoke `super.visitFieldDeclaration(node)` so that the initializer will be visited because late instance fields can reference other instance members.